fmn/formatters.py

```diff
@@ -0,0 +1,426 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of the FMN project.
+# Copyright (C) 2017 Red Hat, Inc.
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+"""
+Formatters for message backends.
+
+These take raw fedmsgs and produce human-readable notifications as text. This
+text is then passed to a delivery backend and sent to the user.
+"""
+from __future__ import absolute_import
+
+import logging
+import json
+import uuid
+import datetime
+import time
+import email as email_module
+
+from kitchen.text.converters import to_bytes, to_unicode
+import arrow
+import requests
+import six
+import pytz
+import fedmsg.meta
+
+from . import config
+
+_log = logging.getLogger(__name__)
+
+MIRC_COLORS = {
+    "white": 0,
+    "black": 1,
+    "blue": 2,
+    "green": 3,
+    "red": 4,
+    "brown": 5,
+    "purple": 6,
+    "orange": 7,
+    "yellow": 8,
+    "light green": 9,
+    "teal": 10,
+    "light cyan": 11,
+    "light blue": 12,
+    "pink": 13,
+    "grey": 14,
+    "light grey": 15,
+}
+
+
+def shorten(link):
+    """
+    Attempt to shorten a link using a shortening service.
+
+    Args:
+        link (str): The link to shorten.
+
+    Returns:
+        str: The link. If the shortener service was unreachable, the original link is returned.
+    """
+    if not link:
+        return ''
+    try:
+        response = requests.get('http://da.gd/s', params=dict(url=link))
```

Comment:
Should we put a timeout on this call? Also, should we use a requests session?

Reply:
This got moved as part of a refactor. I added the timeout. It should use a session, but some care will need to be taken with that since multiprocessing is in the mix here.